language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/codeaction/BallerinaCodeActionProviderFactory.java

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+package org.ballerinalang.langserver.codeaction;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+/**
+ * Represents the Code Action provider factory.
+ *
+ * @since 1.1.0
+ */
+public class BallerinaCodeActionProviderFactory {
+    private static Map<CodeActionNodeType, List<BallerinaCodeActionProvider>> nodeBasedProviders;
+    private static List<BallerinaCodeActionProvider> diagnosticsBasedProviders;
+    private static BallerinaCodeActionProviderFactory codeActionProviderInstance;
+    private ServiceLoader<BallerinaCodeActionProvider> serviceLoader;
+    private boolean isInitialized = false;
+
+    private BallerinaCodeActionProviderFactory() {
+        serviceLoader = ServiceLoader.load(BallerinaCodeActionProvider.class);
+    }
+
+    static {
+        codeActionProviderInstance = new BallerinaCodeActionProviderFactory();
+    }
+
+    /**
+     * @return code action provider factory instance
+     */
+    public static BallerinaCodeActionProviderFactory getInstance() {
+        return codeActionProviderInstance;
+    }
+
+    /**
+     *
+     * @return node based providers
+     */
+    public static Map<CodeActionNodeType, List<BallerinaCodeActionProvider>> getNodeBasedProviders() {
+        return nodeBasedProviders;
+    }
+
+    public static void setNodeBasedProviders(
+            Map<CodeActionNodeType, List<BallerinaCodeActionProvider>> nodeBasedProviders) {
+        BallerinaCodeActionProviderFactory.nodeBasedProviders = nodeBasedProviders;
+    }
+
+    /**
+     *
+     * @return diagnostic based providers
+     */
+    public static List<BallerinaCodeActionProvider> getDiagnosticsBasedProviders() {
+        return diagnosticsBasedProviders;
+    }
+
+    public static void setDiagnosticsBasedProviders(
+            List<BallerinaCodeActionProvider> diagnosticsBasedProviders) {
+        BallerinaCodeActionProviderFactory.diagnosticsBasedProviders = diagnosticsBasedProviders;
+    }
+
+    /**
+     * Initialize the code action provider factory Adding all node based providers in to the Hash Map and diagnostic
+     * based providers in to the List.
+     */
+    public void initiate() {
+        if (isInitialized) {
+            return;
+        }
+        BallerinaCodeActionProviderFactory.setNodeBasedProviders(new HashMap<>());
+        BallerinaCodeActionProviderFactory.setDiagnosticsBasedProviders(new ArrayList<>());
+        for (CodeActionNodeType nodeType : CodeActionNodeType.values()) {
+            BallerinaCodeActionProviderFactory.getNodeBasedProviders().put(nodeType, new ArrayList<>());
+        }
+        for (BallerinaCodeActionProvider codeAction : serviceLoader) {
+            if (codeAction.isNodeBased()) {
+                for (CodeActionNodeType nodeType : codeAction.getCodeActionNodeTypes()) {
+
+                    if (nodeType.equals(CodeActionNodeType.FUNCTION)) {
```

Comment:
Use `==` over `.equals` to compare enums. It is faster and safer. eg. `nodeType == CodeActionNodeType.FUNCTION`

Reply:
fixed